### DISCUSSION TITLE
Inject[Greeter] vs. inject.me()

### DISCUSSION THREAD

**Discussion by pauleveritt** [Finistere/antidote#59] [General]
I'm working again on examples. I'm covering different spelling for injector.

In the [injection example](https://antidote.readthedocs.io/en/stable/tutorial/2_injection.html) you cover the different spellings for injection. I'm curious the difference between `greeter: Greeting = inject.me()` and `greeting: Inject[Greeter]`.

I know the latter can't be used with other markers such as `Optional`. Are there any other differences?

`Inject[Greeter]` is shorter and *might* be more normal to look at. For example, flake8 out-of-the-box complains about `inject.me()`.

I'm considering changing my examples to `Inject[Greeter]` but that differs from official docs.

**Comment by Finistere** [comment]
What kind of error did you have with flake8? I never had any with `inject.me()`. :thinking: 

The main difference is the one you've seen in #60. It just works better with static type checking. It also has better syntax to retrieve implementations for a given interface currently:

```python
def f(a: A = inject.me(qualified_by=X)):
    ...

def g(a: Annotated[A, Get(ImplementationsOf(A).single(qualified_by=X))]):
    ...
```

The V2 I'm working on removes the latter difference though, objects used for annotation and default are the same and treated in the same way. Moreover, they're open for extension. Here's a bit simplified extract:

```python
# implemented by inject.get, lazy functions, constants
class Dependency(Generic[Out]):
    def __antidote_dependency__(self) -> object:
        return self

# implemented by inject.me() and used in Inject annotation
class ParameterDependency:
    def __antidote_parameter_dependency__(
        self, *, name: str, type_hint: object, type_hint_with_extras: object
    ) -> Dependency[Any]:
        raise NotImplementedError()
```

To be honest, the best way to use antidote is definitely using default parameters as it's the only one providing proper static type checking. Now that said, previous injection styles have been there before and the purpose of the core part is to be flexible. So I intend to keep them as I don't see a good reason to remove them. I probably should remove them from the README and provide an explanation of their shortcomings though.

**Comment by pauleveritt** [comment]
Agreed, changing the docs to show these as kind of legacy is a good answer.

**Reply by Finistere** [reply]
Thanks for pointing it out!
